src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -858,21 +826,20 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
         'by a bug in React. Please file an issue.',
     );
     isPerformingWork = true;
-    const isPerformingDeferredWork = !!deadline;
 
     // This outer loop exists so that we can restart the work loop after
     // catching an error. It also lets us flush Task work at the end of a
     // deferred batch.
-    while (priorityLevel !== NoWork && !fatalError) {
-      invariant(
-        deadline !== null || priorityLevel < HighPriority,
-        'Cannot perform deferred work without a deadline. This error is ' +
-          'likely caused by a bug in React. Please file an issue.',
-      );
-
+    let keepPerformingWork = true;
+    let hasRemainingAsyncWork = false;
+    while (keepPerformingWork && fatalError === null) {
```

Comment:
What's the intention here with all the possible states (9) these can be in? I have trouble following the extra flags. The flags are indicative of a code smell. Can you refactor this in terms of control flow?